lib/grppcaut.gi

```diff
@@ -869,12 +907,319 @@ end;
+    # mixer
+    m:=IdentityMat(n,f);
+    m[l+1][1]:=One(f);
+    Add(stb,m);
+    stb:=List(stb,x->x^act); # base change
+    stb:=Group(stb);
+    SetSize(stb,Size(GL(l,f))*Size(GL(n-l,f))*Size(f)^(l*(n-l)));
+    Info(InfoMatOrb,2,"reduced full GL by ",Size(G)/Size(stb));
+    return stb;
+  fi;
+  if not OrbitIsLonger(GeneratorsOfGroup(G),bas,OnSubspacesByCanonicalBasis,
+	   Size(f)^QuoInt(Length(bas[1]),2)) then
+    #just plain
+    act:=OnSubspacesByCanonicalBasis;
+    xset:=Orbit(G,bas,act);
+    stb:=[Position(xset,bas)];
+  else
+    act:=OnLines;
+    stb:=NormedVectors(VectorSpace(f,bas));
+    xset:=Union(Orbits(G,stb,act));
+    stb:=Set(List(stb,x->Position(xset,x)));
+  fi;
+  hom:=ActionHomomorphism(G,xset,act,"surjective");
+  act:=PreImage(hom,Stabilizer(Image(hom),stb,OnSets));
+  Info(InfoMatOrb,2,"reduce ",Size(G), " to ",Size(act));
+  return act;
+end);
+
+# construct subgroup of GL that stabilizes the spaces given and fixes teh
+# listed spaceorbits.
+
+# auxillary
+RedmatSpanningIndices:=function(gens)
+local bas,n,one,new,a,b,g;
+  n:=Length(gens[1]);
+  one:=One(gens[1]);
+  new:=[];
+  bas:=[];
+  while Length(bas)<n do
+    a:=First(one,x->Length(bas)=0 or SolutionMat(bas,x)=fail);
+    Add(new,Position(one,a));
+    Add(bas,a);
+    # spin
+    for b in bas do
+      for g in gens do
+        a:=b*g;
+	if SolutionMat(bas,a)=fail then
+	  Add(bas,a);
+	fi;
+      od;
+    od;
+  od;
+  return new;
+end;
+
+SpaceAndOrbitStabilizer:=function(n,field,spaces,osporb)
+local spaceincl,outvecs,l,sub,yet,i,j,k,s,t,new,incl,min,rans,sofar,done,
+      gens,one,spl,ngens,m,sz,a,sporb,notyet;
+
+  spaceincl:=function(big,small)
+    return RankMat(Concatenation(big,small))=Length(big);
+  end;
+
+  outvecs:=function(space,new)
+    return BaseSteinitzVectors(new,space).factorspace;
+  end;
+
+  one:=IdentityMat(n,field);
+  sub:=[]; # space so far
+  spaces:=Unique(List(spaces,x->OnSubspacesByCanonicalBasis(x,one)));
+
+  SortBy(spaces,Length);
+  if not ForAny(spaces,x->Length(x)=n) then
+    Add(spaces,List(one,ShallowCopy));
+  fi;
+
+  osporb:=List(osporb,x->List(x,x->OnSubspacesByCanonicalBasis(x,one)));
+  sporb:=List(osporb,ShallowCopy);
+  l:=Length(spaces);
+
+  # inclusion relation, tests
+  repeat
+    yet:=0;
+    done:=[0];
+    incl:=List([1..l],x->[0,x]);
+    new:=[];
+    for i in [1..l] do
+      for j in [i+1..l] do
+	if spaceincl(spaces[j],spaces[i]) then
+	  AddSet(incl,[i,j]);
+	else
+	  # check for meet/join closed
+	  s:=SumIntersectionMat(spaces[i],spaces[j]);
+	  for k in s do
+	    if Length(k)>0 and not ForAny(spaces,x->Length(x)=Length(k) and
+	      RankMat(Concatenation(x,k))=Length(x)) then
+	      Add(new,k);
+	    fi;
+	  od;
+	fi;
+      od;
+    od; 
+    if Length(new)>0 then
+      new:=List(new,x->OnSubspacesByCanonicalBasis(x,one)); # canonize
+      spaces:=Set(Concatenation(spaces,new));
+      l:=Length(spaces);
+      SortBy(spaces,Length);
+    fi;
+  until Length(new)=0;
+
```

Comment:
Restarting like this can throw away a lot of perfectly fine information. In a bad case I have, we start out with about 4000 spaces, which is bad enough; but it then grows to NNN spaces, and starts over.

Why not instead have a workqueue: Initially, all spaces we in this workqueue, and `spaces` is empty. Instead of iterating over `spaces,` you iterate over the queue, and gradually add its entries to `spaces` as well as their sum+intersection with all elements already in `spaces` at that point. This way, you don't have to restart the loop.

The queue should perhaps also be grouped by dimensions (just as I suggested for `spaces`, so that we can deal with 1-dimensional spaces first, if we want to, as those are easier to handle